libs/cli/src/main/java/org/elasticsearch/cli/Terminal.java

```diff
@@ -339,4 +363,54 @@ public OutputStream getOutputStream() {
             return System.out;
         }
     }
+
+    /** A line based OutputStream wrapping this Terminal's println, not thread-safe! */
+    private class LineOutputStream extends OutputStream {
+        static final int DEFAULT_BUFFER_LENGTH = 1024;
+        static final int MAX_BUFFER_LENGTH = DEFAULT_BUFFER_LENGTH * 8;
+
+        private final Charset charset;
+        private byte[] bytes = new byte[DEFAULT_BUFFER_LENGTH];
+        private int count = 0;
+
+        LineOutputStream(Charset charset) {
+            this.charset = charset;
+        }
+
+        @Override
+        public void write(int b) {
+            if (b == 0) return;
+            if (b == '\n') {
+                flushLine(false);
+                return;
+            }
+            if (count == bytes.length) {
+                if (count >= MAX_BUFFER_LENGTH) {
+                    flushLine(false);
+                } else {
+                    bytes = Arrays.copyOf(bytes, 2 * bytes.length);
+                }
+            }
+            bytes[count++] = (byte) b;
+        }
+
+        private void flushLine(boolean skipEmpty) {
+            if (count > 0 && bytes[count - 1] == '\r') {
+                --count; // drop CR on windows as well
+            }
+            if (skipEmpty && count == 0) {
+                return;
+            }
+            println(count > 0 ? new String(bytes, 0, count, charset) : "");
```

Comment:
nit: if we flush in the case of hitting the max buffer length, then println might not be correct? Perhaps we could pass through whether a new line was actually encountered and decide whether to println or print?

Reply:
Done